gnuradio-runtime/lib/hier_block2_detail.cc

```diff
@@ -996,5 +1017,70 @@ void hier_block2_detail::reset_hier_blocks_parent()
     }
 }
 
+std::set<basic_block_sptr> hier_block2_detail::all_blocks()
+{
+    std::set<basic_block_sptr> found_blocks;
+    std::deque<hier_block2_sptr> queued_blocks;
+
+    // Get the blocks from our own flow graph. These can contain hier blocks
```

Comment:
Would it make sense to eliminate this part of the code by adding this block to `queued_blocks` at the beginning and then going directly to the `while (!queued_blocks.empty())` loop?

Reply:
ah wait, no. The problem was that for the current `hier_block2_detail`, I don't have an sptr to the owning hier block, only a raw pointer.

I don't want to change `queued_blocks` to be `container<hier_block2*>`, because that would break the thread-safety of enumerating blocks (the result would still be nonsensical if someone was changing the internal block connections while I'm doing this, but at least it would still hold shared ownership of the contained blocks, which hence continue to live, and thus, this walk would not segfault).

The solution to that might be documenting you shouldn't change flowgraphs while you're setting affinity or log level, but I'd rather keep it safe and non-surprising; but as me forgetting this since Sunday highlights this should be commented on in the code, for it not to fall prey to future refactors.

Reply:
That makes sense. I read the new comment and it seems good.